### PR TITLE
[WIP] Tree Sitter Layer

### PR DESCRIPTION
**Pull request by sdwolfz**
There is much work to be done here, but this is the initial support for tree-sitter. You can enable it with...
```elisp
(tree-sitter
  :variables
    tree-sitter-toggle 'global)
```
... in your `dotspacemacs-configuration-layers`.

It seems to work, and appears to load files faster, but the colors are... weird, at least for me with the monokai theme, I'll probably have to deal with that upstream.

For the final PR I'll make it a default layer and have it toggle on demand for the supported languages. When I'm convinced it's mature enough I'll have it globally by default.

### CODE REVIEW

File: layers/+tools/tree-sitter/packages.el [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+(defconst tree-sitter-packages
+  '(
+    tree-sitter
+    tree-sitter-langs))
+
+(defun tree-sitter/init-tree-sitter ()
+  (use-package tree-sitter
+    :defer (eq tree-sitter-toggle 'on-demand)
+    :init
+    (if (eq tree-sitter-toggle 'global)
```
Comment (lebensterben) [inline layers/+tools/tree-sitter/packages.el:32]: if -> when

**Comment by duianto** [comment]
`tree-sitter-langs` seems to assume that it contains a git directory,
the following messages appears on every startup:
```
[tree-sitter-langs] Running (git rev-parse --show-toplevel) in c:/Users/username/AppData/Roaming/.emacs.d/elpa/27.2/develop/tree-sitter-langs-20210918.1621/
The directory c:/Users/username/AppData/Roaming/.emacs.d/elpa/27.2/develop/tree-sitter-langs-20210918.1621/ doesn’t seem to be a git working dir. Grammar-building functions will not work.
tree-sitter-langs: Grammar bundle v0.10.7 was already installed; skipped
Clearing removed files...done
Processing modified files...done
```

**Comment by sdwolfz** [comment]
> `tree-sitter-langs` seems to assume that it contains a git directory, the following messages appears on every startup:
> 
> ```
> [tree-sitter-langs] Running (git rev-parse --show-toplevel) in c:/Users/username/AppData/Roaming/.emacs.d/elpa/27.2/develop/tree-sitter-langs-20210918.1621/
> The directory c:/Users/username/AppData/Roaming/.emacs.d/elpa/27.2/develop/tree-sitter-langs-20210918.1621/ doesn’t seem to be a git working dir. Grammar-building functions will not work.
> tree-sitter-langs: Grammar bundle v0.10.7 was already installed; skipped
> Clearing removed files...done
> Processing modified files...done
> ```

I've noticed this as well, going to dig through it and see why.

File: layers/+lang/ruby/packages.el [hunk 2]
```diff
@@ -316,6 +316,14 @@
       (add-hook 'ruby-mode-hook #'spacemacs//ruby-setup-backend)
       (add-hook 'ruby-mode-local-vars-hook
                 #'spacemacs/ruby-maybe-highlight-debugger-keywords))
+      ;; Tree-sitter on demand.
+      (configuration-layer/with-dynamic-modules
+        (when (and
+                (configuration-layer/layer-used-p 'tree-sitter)
+                (eq tree-sitter-toggle 'on-demand)
+                (eq ruby-syntax-backend 'tree-sitter))
+          (message "RUBY AND EMACS SITTING IN A TREE!")
```
Comment (robbyoconnor) [inline layers/+lang/ruby/packages.el:325]: Oh my god, I love it!

**Pull request closed without merging**